astropy/utils/introspection.py

```diff
@@ -2,20 +2,39 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """Functions related to Python runtime introspection."""
 
-
+import collections
 import inspect
-import re
 import os
 import sys
 import types
 import importlib
+from importlib import metadata
+from packaging.version import Version
+
+from astropy.utils.decorators import deprecated_renamed_argument
 
 __all__ = ['resolve_name', 'minversion', 'find_current_module',
            'isinstancemethod']
 
-
 __doctest_skip__ = ['find_current_module']
 
+if sys.version_info[:2] >= (3, 10):
+    from importlib.metadata import packages_distributions
+else:
+    def packages_distributions():
+        """
+        Return a mapping of top-level packages to their distributions.
+        Note: copied from https://github.com/python/importlib_metadata/pull/287
+        """
+        pkg_to_dist = collections.defaultdict(list)
+        for dist in metadata.distributions():
+            for pkg in (dist.read_text('top_level.txt') or '').split():
+                pkg_to_dist[pkg].append(dist.metadata['Name'])
+        return dict(pkg_to_dist)
+
+
+_PKG_TO_DIST_NAMES = {}
```

Comment:
The original goal was to simplify things but is this really simplifying? Should we wait till minversion is 3.10? We just dropped 3.7, so that will be a few years down the road. 💭

Reply:
We need to get rid of distutils (https://github.com/astropy/astropy/pull/11962#discussion_r673161866).